[quant][graphmode][fix] dequantize propagation for {add/mul}_scalar

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40596 [quant][graphmode][fix] dequantize propagation for {add/mul}_scalar**

Summary:
Previously the fusion patterns for {add/mul}_scalar is inconsistent since the op pattern
produces a non-quantized tensor and the op replacement graph produces a quantized tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22251072](https://our.internmc.facebook.com/intern/diff/D22251072)

jerryzh168 committed:
[quant][graphmode][fix] dequantize propagation for {add/mul}_scalar

Summary:
Previously the fusion patterns for {add/mul}_scalar is inconsistent since the op pattern
produces a non-quantized tensor and the op replacement graph produces a quantized tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2272,6 +2272,10 @@ def __init__(self):
 
             def forward(self, x):
                 x = self.conv(x)
+                # add_scalar
```
I thought this was already supported. What was missing?

jerryzh168 replied:
the fusion pattern is inconsistent

jerryzh168 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -504,29 +527,15 @@ graph(%a_quant, %b_quant, %alpha, %scale, %zero_point, %dtype):
          return (%r) )";
 
-  // quantized::add_scalar
-  std::string add_scalar = R"(
-graph(%a_quant, %b_scalar, %alpha):
-         %a_dequant = aten::dequantize(%a_quant)
-         %r = aten::add(%a_dequant, %b_scalar, %alpha)
-         return (%r) )";
-
-  std::string quantized_add_scalar = R"(
-graph(%a_quant, %b_scalar, %alpha):
-         %r = quantized::add_scalar(%a_quant, %b_scalar)
-         return (%r) )";
```
@raghuramank100 major change is this

jerryzh168 replied:
original pattern produces a float tensor and quantized pattern produces a quantized tensor

raghuramank100 replied:
So is it correct to say that the new pattern is: deq->add-> = quantize_add->deq?

jerryzh168 replied:
new pattern is:
dq - aten::add -quant --> quantized::add_scalar

jerryzh168 committed:
Update on "[quant][graphmode][fix] dequantize propagation for {add/mul}_scalar"


Summary:
Previously the fusion patterns for {add/mul}_scalar is inconsistent since the op pattern
produces a non-quantized tensor and the op replacement graph produces a quantized tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] dequantize propagation for {add/mul}_scalar"


Summary:
Previously the fusion patterns for {add/mul}_scalar is inconsistent since the op pattern
produces a non-quantized tensor and the op replacement graph produces a quantized tensor

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22251072](https://our.internmc.facebook.com/intern/diff/D22251072)

[ghstack-poisoned]
